### PR TITLE
nebula bench2.0

### PR DESCRIPTION
**Pull request by tom-chensf**
1、LDBC data import
2、Jmeter perf test

### CODE REVIEW

File: util/ldbc_go_step/src/main/java/vesoft/ldbc_go_step.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+import org.apache.jmeter.samplers.SampleResult;
+import org.apache.jmeter.samplers.SampleResult;
+import org.apache.jmeter.assertions.AssertionResult;
+
+
+/**
+ * LDBC Go Step
+ *
+ */
+public class ldbc_go_step implements JavaSamplerClient
```
Comment (CPWstatic) [inline util/ldbc_go_step/src/main/java/vesoft/ldbc_go_step.java:27]: LdbcGoStep

File: util/ldbc_go_step/src/main/java/vesoft/ldbc_go_step.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        arguments.addArgument("user","root");
+        arguments.addArgument("pwd","nebula");
+        arguments.addArgument("space","");
+        arguments.addArgument("nGQL","");
+	arguments.addArgument("person","");
+        return arguments;
+    }
+
+
+   public synchronized void init_nebula_pool(String hosts, int maxconn)
```
Comment (CPWstatic) [inline util/ldbc_go_step/src/main/java/vesoft/ldbc_go_step.java:49]: initNebulaPool

Comment (tom-chensf) [reply util/ldbc_go_step/src/main/java/vesoft/ldbc_go_step.java]: OK

File: util/ldbc_go_step/src/main/java/vesoft/ldbc_go_step.java (hunk 1 above)
Comment (tom-chensf) [reply util/ldbc_go_step/src/main/java/vesoft/ldbc_go_step.java]: OK

File: README.md [hunk 3]
```diff
@@ -1,190 +1,99 @@
...
-### Query Benchmark
-```shell
-$ ./nb --bench-config=ldbc/query/3-hop-knows.json
-```
+# 1、Download nebula-bench 
+    
+       https://github.com/vesoft-inc/nebula-bench.git
+ 
+# 2、Download Jmeter
+       cd nebula-bench &&  wget https://mirror.bit.edu.cn/apache//jmeter/binaries/apache-jmeter-5.4.zip  &&  unzip apache-jmeter-5.4.zip
```
Comment (laura-ding) [inline README.md:6]: why download the jmeter in the dir of nebula-bench. it will contaminate the source directory.

File: util/LdbcGoStep/pom.xml [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+      </dependency>
+      <dependency>
+          <groupId>org.slf4j</groupId>
+          <artifactId>slf4j-api</artifactId>
+          <version>1.7.25</version>
+      </dependency>
+      <dependency>
+          <groupId>com.vesoft</groupId>
+          <artifactId>client</artifactId>
+          <version>2.0.0-beta</version>
```
Comment (laura-ding) [inline util/LdbcGoStep/pom.xml:24]: update the version, the interface with the nebula-graph has change

File: util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java [hunk 5]
```diff
@@ -0,0 +1,143 @@
+package com.vesoft;
```
Comment (laura-ding) [inline util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java:1]: add head

```
/* Copyright (c) 2021 vesoft inc. All rights reserved.
 *
 * This source code is licensed under Apache 2.0 License,
 * attached with Common Clause Condition 1.0, found in the LICENSES directory.
 */

```

File: util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+public class LdbcGoStep implements JavaSamplerClient {
+    private static final Logger log = LoggerFactory.getLogger(LdbcGoStep.class);
+    private NebulaPool pool = null;
+    private Session session = null;
+
+
+    @Override
+    public Arguments getDefaultParameters() {
+        Arguments arguments = new Arguments();
+        arguments.addArgument("hosts", "127.0.0.1:3699");
```
Comment (laura-ding) [inline util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java:36]: modify default port to 9669

File: util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+        arguments.addArgument("space", "");
+        arguments.addArgument("nGQL", "");
+        arguments.addArgument("person", "");
+        return arguments;
+    }
+
+
+    public void initNebulaPool(String hosts, int maxconn, int id) {
+        pool = new NebulaPool();
+        assert pool != null;
```
Comment (laura-ding) [inline util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java:49]: it's redundant

File: util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+            nebulaPoolConfig.setMaxConnSize(maxconn);
+            List<String> host_list = new ArrayList<String>(Arrays.asList(hosts.split(",")));
+            assert host_list != null;
+            String host = host_list.get(id%host_list.size());
+            String[] splits = host.split(":");
+            addresses.add(new HostAddress(splits[0], Integer.parseInt(splits[1])));
+            boolean init = pool.init(addresses, nebulaPoolConfig);
+            assert init == true;
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (laura-ding) [inline util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java:62]: you need to exit the process, not only print the stacktrace

File: util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        initNebulaPool(hosts, maxconn, id);
+        try {
+            session = pool.getSession(user, pwd, false);
+            assert session != null;
+            String use_space = "use " + space + ";";
+            ResultSet resp = null;
+            resp = session.execute(use_space);
+            assert resp.isSucceeded();
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (laura-ding) [inline util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java:86]: ditto

File: util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+        }
+    }
+
+
+    @Override
+    public SampleResult runTest(JavaSamplerContext javaSamplerContext) {
+        System.out.print("Perf runTest:" + Thread.currentThread().getName() + "\n");
```
Comment (laura-ding) [inline util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java:93]: `println` with the `\n`,

File: util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+        System.out.print("Pert test end!\n");
+        assert session != null;
+        session.release();
+        if (pool.getActiveConnNum() == 0) {
+            pool.close();
+        }
+    }
+
+    public static void main(String[] args) {
+        System.out.println("Hello World!");
```
Comment (laura-ding) [inline util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java:141]: modify here

File: util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+            result.setSuccessful(true);
+        }
+        return result;
+    }
+
+
+    @Override
+    public void teardownTest(JavaSamplerContext javaSamplerContext) {
+        System.out.print("Pert test end!\n");
+        assert session != null;
```
Comment (laura-ding) [inline util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java:133]:
```
if (session != null) {
    session.release();
}
```

File: util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+            List<HostAddress> addresses = new ArrayList();
+            NebulaPoolConfig nebulaPoolConfig = new NebulaPoolConfig();
+            nebulaPoolConfig.setMaxConnSize(maxconn);
+            List<String> host_list = new ArrayList<String>(Arrays.asList(hosts.split(",")));
+            assert host_list != null;
+            String host = host_list.get(id%host_list.size());
+            String[] splits = host.split(":");
+            addresses.add(new HostAddress(splits[0], Integer.parseInt(splits[1])));
+            boolean init = pool.init(addresses, nebulaPoolConfig);
+            assert init == true;
```
Comment (laura-ding) [inline util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java:60]: not use assert, add judge and add log to print

File: README.md (hunk 3 above)
Comment (tom-chensf) [reply README.md:6]: This is  a example,user can change dir,I change it.

File: util/LdbcGoStep/pom.xml (hunk 4 above)
Comment (tom-chensf) [reply util/LdbcGoStep/pom.xml:24]: I change it to 2.0.0-rc1

File: util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java (hunk 5 above)
Comment (tom-chensf) [reply util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java:1]: ok

File: util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java (hunk 6 above)
Comment (tom-chensf) [reply util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java:36]: ok

File: util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java (hunk 7 above)
Comment (tom-chensf) [reply util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java:49]: ok

File: util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java (hunk 10 above)
Comment (tom-chensf) [reply util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java:93]: ok

File: util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java (hunk 12 above)
Comment (tom-chensf) [reply util/LdbcGoStep/src/main/java/vesoft/LdbcGoStep.java:133]: ok

**Review by monadbobo — APPROVED** [review]
LGTM.

File: README.md [hunk 14]
```diff
@@ -1,190 +1,101 @@
...
+        ./nebula-importer --config  ldbc_vid_int.yaml
+ 
+#  4、Perf test
+##     Mvn package
+       cd util/LdbcGoStep/ && mvn package
+      
+      
+##     Put  jar and jmx file to Jmeter 
+       LdbcGoStep-2-jar-with-dependencies.jar: put to  apache-jmeter-5.4/lib/ext
+       nebula-bench/ldbc/jmx/go_step.jmx : put to apache-jmeter-5.4/
```
Comment (CPWstatic) [inline README.md:57]: Would you please merge this part to the setup.sh?

File: README.md [hunk 15]
```diff
@@ -1,190 +1,101 @@
...
+
+
+##    Run Jmeter
+       cd apache-jmeter-5.4 
+       perf test:
+        ./bin/jmeter.sh -n -t go_step.jmx  -l go_step.jtl -j go_step.log
+       report(jmeter):
+        ./bin/jmeter.sh -g go_step.jtl  -o go_step
+       report(statictis):
+        python3 ldbc/scripts/statistics.py go_step.jtl
```
Comment (CPWstatic) [inline README.md:96]: Would you please merge this part into a script? That would be helpful much more to the users.

**Review by CPWstatic — APPROVED** [review]

**Pull request merged**